core/system/src/export_table.c

```diff
@@ -14,7 +14,105 @@
+#endif
+
+/* Per thread we store the pointer to the allocator and the process id that
+ * this thread belongs to. */
+typedef struct {
+    void * allocator;
+    int process_id;
+} UvisorThreadContext;
+
+static UvisorThreadContext thread[UVISOR_EXPORT_TABLE_THREADS_MAX_COUNT];
+
+
+static int thread_ctx_valid(UvisorThreadContext * context)
+{
+    /* Check if context pointer points into the array. */
+    if ((uint32_t) context < (uint32_t) &thread ||
+        ((uint32_t) &thread + sizeof(thread)) <= (uint32_t) context) {
+        return 0;
+    }
+    /* Check if the context is aligned exactly to a context. */
+    return (((uint32_t) context - (uint32_t) thread) % sizeof(UvisorThreadContext)) == 0;
+}
+
+static void * thread_create(int id, void * c)
+{
+    (void) id;
+    UvisorThreadContext * context = c;
+    const UvisorBoxIndex * const index =
+            (UvisorBoxIndex * const) *(__uvisor_config.uvisor_box_context);
+    /* Search for a free slot in the tasks meta data. */
+    uint32_t ii = 0;
+    for (; ii < UVISOR_EXPORT_TABLE_THREADS_MAX_COUNT; ii++) {
+        if (thread[ii].allocator == NULL)
+            break;
+    }
+    if (ii < UVISOR_EXPORT_TABLE_THREADS_MAX_COUNT) {
+        /* Remember the process id for this thread. */
+        thread[ii].process_id = g_active_box;
+        /* Fall back to the process heap if ctx is NULL. */
+        thread[ii].allocator = context ? context : index->box_heap;
+        return &thread[ii];
+    }
+    return context;
+}
+
+static void thread_destroy(void * c)
+{
+    UvisorThreadContext * context = c;
+    if (context == NULL) return;
+
+    /* Only if TID is valid and destruction status is zero. */
+    if (thread_ctx_valid(context) && context->allocator &&
+        (context->process_id == g_active_box)) {
+        /* Release this slot. */
+        context->allocator = NULL;
+    } else {
+        HALT_ERROR(SANITY_CHECK_FAILED,
+            "thread context (%08x) is invalid!\n",
+            context);
+    }
+}
+
+static void thread_switch(void * c)
+{
+    UvisorThreadContext * context = c;
+    UvisorBoxIndex * index;
+    if (context == NULL) return;
+
+    /* Only if TID is valid and the slot is used */
+    if (!thread_ctx_valid(context) || context->allocator == NULL) {
+        HALT_ERROR(SANITY_CHECK_FAILED,
+            "thread context (%08x) is invalid!\n",
+            context);
+        return;
+    }
+    /* If the thread is inside another process, switch into it. */
+    if (context->process_id != g_active_box) {
+        context_switch_in(CONTEXT_SWITCH_UNBOUND_THREAD, context->process_id, 0, 0);
+    }
+    /* Copy the thread allocator into the (new) box index. */
+    index = (UvisorBoxIndex *) *(__uvisor_config.uvisor_box_context);
+    if (context->allocator) {
+        /* If the active_heap is NULL, then the process heap needs to be
+         * initialized yet. The initializer sets the active heap itself. */
+        if (index->active_heap) {
```

Comment:
I see what you are doing. I don't like very much overloading of functionality into a single variable. That smells like trouble when extending code in future. How about an ->initialized flag instead ?